File: springboot/src/main/java/com/mcp/server/handler/McpHandler.java
package com.mcp.server.handler;

import com.mcp.server.tools.WeatherTool;
import com.mcp.server.tools.TemperatureTool;
import com.mcp.server.tools.AirQualityTool;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

import java.util.*;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor
public class McpHandler {

    private final WeatherTool weatherTool;
    private final TemperatureTool temperatureTool;
    private final AirQualityTool airQualityTool;
    
    private List<Object> getRegisteredTools() {
        List<Object> tools = new ArrayList<>();
        tools.add(weatherTool);
        tools.add(temperatureTool);
        tools.add(airQualityTool);
        return tools;
    }

    public Mono<Map<String, Object>> handleRequest(Map<String, Object> request) {
        String method = (String) request.get("method");
        Object id = request.get("id");
        Map<String, Object> params = (Map<String, Object>) request.get("params");

        if ("initialize".equals(method)) {
            return Mono.just(createResponse(id, Map.of(
                "protocolVersion", "0.1.0",
                "serverInfo", Map.of("name", "springboot-mcp-server", "version", "0.1.0"),
                "capabilities", Map.of("tools", Map.of())
            )));
        } else if ("tools/list".equals(method)) {
            List<Map<String, Object>> toolsList = getRegisteredTools().stream()
                .map(tool -> {
                    if (tool instanceof WeatherTool) {
                        WeatherTool wt = (WeatherTool) tool;
                        return Map.of(
                            "name", wt.getName(),
                            "description", wt.getDescription(),
                            "inputSchema", wt.getInputSchema()
                        );
                    } else if (tool instanceof TemperatureTool) {
                        TemperatureTool tt = (TemperatureTool) tool;
                        return Map.of(
                            "name", tt.getName(),
                            "description", tt.getDescription(),
                            "inputSchema", tt.getInputSchema()
                        );
                    } else if (tool instanceof AirQualityTool) {
                        AirQualityTool aqt = (AirQualityTool) tool;
                        return Map.of(
                            "name", aqt.getName(),
                            "description", aqt.getDescription(),
                            "inputSchema", aqt.getInputSchema()
                        );
                    }
                    return Map.<String, Object>of();
                })
                .collect(Collectors.toList());
            
            return Mono.just(createResponse(id, Map.of("tools", toolsList)));
        } else if ("tools/call".equals(method)) {
            if (params == null || !params.containsKey("name")) {
                return Mono.just(createErrorResponse(id, -32602, "Invalid params: 'name' is required"));
            }
            
            String toolName = (String) params.get("name");
            Map<String, Object> arguments = (Map<String, Object>) params.getOrDefault("arguments", Map.of());
            
            return executeToolByName(toolName, arguments)
                .map(result -> createResponse(id, Map.of(
                    "content", List.of(Map.of("type", "text", "text", result))
                )))
                .onErrorResume(e -> Mono.just(createErrorResponse(id, -32000, "Tool execution error: " + e.getMessage())));
        } else {
            return Mono.just(createErrorResponse(id, -32601, "Method not found"));
        }
    }
    
    private Mono<String> executeToolByName(String toolName, Map<String, Object> arguments) {
        for (Object tool : getRegisteredTools()) {
            if (tool instanceof WeatherTool && "get_weather".equals(toolName)) {
                return ((WeatherTool) tool).execute(arguments);
            } else if (tool instanceof TemperatureTool && "get_temperature".equals(toolName)) {
                return ((TemperatureTool) tool).execute(arguments);
            } else if (tool instanceof AirQualityTool && "get_air_quality".equals(toolName)) {
                return ((AirQualityTool) tool).execute(arguments);
            }
        }
        return Mono.error(new RuntimeException("Tool not found: " + toolName));
    }

    private Map<String, Object> createResponse(Object id, Object result) {
        Map<String, Object> response = new HashMap<>();
        response.put("jsonrpc", "2.0");
        response.put("id", id);
        response.put("result", result);
        return response;
    }

    private Map<String, Object> createErrorResponse(Object id, int code, String message) {
        Map<String, Object> response = new HashMap<>();
        response.put("jsonrpc", "2.0");
        response.put("id", id);
        response.put("error", Map.of("code", code, "message", message));
        return response;
    }
}

File: python/src/mcp_handler.py
from fastapi import APIRouter, Request, HTTPException
from pydantic import BaseModel, Field
from typing import Any, Dict, List, Optional, Union
import logging
from src.tool_registry import tools_registry
# Import tools to register them
import src.tools.weather_tool
import src.tools.temperature_tool
import src.tools.air_quality_tool

router = APIRouter()
logger = logging.getLogger(__name__)

class JsonRpcRequest(BaseModel):
    jsonrpc: str = "2.0"
    method: str
    params: Optional[Union[Dict[str, Any], List[Any]]] = None
    id: Optional[Union[str, int]] = None

class JsonRpcResponse(BaseModel):
    jsonrpc: str = "2.0"
    result: Optional[Any] = None
    error: Optional[Dict[str, Any]] = None
    id: Optional[Union[str, int]] = None

@router.post("/mcp")
async def handle_mcp_request(request: JsonRpcRequest):
    logger.info(f"Received MCP request: {request.method}")
    
    try:
        if request.method == "initialize":
            return JsonRpcResponse(
                id=request.id,
                result={
                    "protocolVersion": "0.1.0",
                    "serverInfo": {
                        "name": "python-mcp-server",
                        "version": "0.1.0"
                    },
                    "capabilities": {
                        "tools": {}
                    }
                }
            )
        
        elif request.method == "tools/list":
            tools_list = [
                {
                    "name": tool["name"],
                    "description": tool["description"],
                    "inputSchema": tool["inputSchema"]
                }
                for tool in tools_registry.values()
            ]
            return JsonRpcResponse(
                id=request.id,
                result={
                    "tools": tools_list
                }
            )
            
        elif request.method == "tools/call":
            if not request.params or "name" not in request.params:
                 return JsonRpcResponse(
                    id=request.id,
                    error={"code": -32602, "message": "Invalid params: 'name' is required"}
                )
            
            tool_name = request.params["name"]
            arguments = request.params.get("arguments", {})
            
            if tool_name not in tools_registry:
                return JsonRpcResponse(
                    id=request.id,
                    error={"code": -32601, "message": f"Tool not found: {tool_name}"}
                )
            
            tool = tools_registry[tool_name]
            try:
                # Execute the tool function
                result = await tool["func"](**arguments)
                return JsonRpcResponse(
                    id=request.id,
                    result={
                        "content": [
                            {
                                "type": "text",
                                "text": str(result)
                            }
                        ]
                    }
                )
            except Exception as e:
                logger.error(f"Error executing tool {tool_name}: {str(e)}")
                return JsonRpcResponse(
                    id=request.id,
                    error={"code": -32000, "message": f"Tool execution error: {str(e)}"}
                )

        else:
            return JsonRpcResponse(
                id=request.id,
                error={"code": -32601, "message": "Method not found"}
            )
            
    except Exception as e:
        logger.error(f"Internal error: {str(e)}")
        return JsonRpcResponse(
            id=request.id,
            error={"code": -32603, "message": "Internal error"}
        )
